treewide: use outputDoc = "devdoc" instead of moveToOutput

Artturin opened the pull request:
this achieves the same effect but in less lines

we cannot control the output location of gi-docgen

example of how the output location is set
https://github.com/frida/json-glib/blob/a8d3ac569bfaf509e2a20b55ee4fd6b89851b8b1/doc/meson.build#L8

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin commented:
```
gnome-builder
gnome.ghex
gnome.eog
evince
devhelp
gdk-pixbuf
gssdp
gtk4
gtksourceview
json-glib
jsonrpc-glib
libadwaita
libgweather
libhandy
libpeas
libportal
librest
librsvg
libsecret
libshumate
pango
fwupd
```

all built

jtojnar commented:
I specifically wanted to move just the developer documentation (directories containing `devhelp2` files). Though some packages incorrectly just move everything in `share/doc`. Ideally, we would have the multi-outputs hook handle it. Some of those packages might not currently have anything else in `share/doc` but that can change – e.g. in the past, there was a push to include README and LICENSE in there.

Artturin committed:
multiple-outputs.sh: move gi-docgen docs in _multioutDocs

Artturin commented:
if we don't want to move the whole dir then we will have to move files ending in these and more
```
$ dir | rev | cut -d. -f1 | rev | sort | uniq
css
devhelp2
html
js
json
png
woff
woff2
```

zseri commented:
As part of a bump of zrythm I'm packaging `libpanel`, should I try to already include this change, or is this too much in-flux?

jtojnar commented:
libpanel is already packaged in #182618

zseri commented:
is there some estimate when #182618 will be merged, or should I just rebase my PR on top of it?

jtojnar commented:
Ideally September 17, since that is when GNOME 43 is scheduled to release. If the Hydra situation resolves before then. But we might need to wait for stragglers.

jtojnar commented on pkgs/build-support/setup-hooks/multiple-outputs.sh:
```diff
@@ -154,6 +154,16 @@ _multioutDocs() {
+        if [ -d "${!output}/share/doc" ]; then
+            while IFS= read -r -d '' dir; do
+                moveToOutput "share/doc/$(basename "$dir")" "${!outputDevdoc}"
+            done <  <(find "${!output}/share/doc" -type f -name "*devhelp*" -exec dirname "{}" \;)
```
This does not seem to print `NUL` delimited paths as expected by `read -d ''`.

jtojnar replied:
Also might be safer to use `*.devhelp2` as the pattern, gi-docgen does not generate older index versions.

jtojnar commented on pkgs/build-support/setup-hooks/multiple-outputs.sh:
```diff
@@ -154,6 +154,16 @@ _multioutDocs() {
+    # move gi-docgen docs to devdoc
+    for output in $(getAllOutputNames); do
+        if [ -d "${!output}/share/doc" ]; then
+            while IFS= read -r -d '' dir; do
```
Maybe use `-d $'\0'` to be explicit.

jtojnar commented on pkgs/build-support/setup-hooks/multiple-outputs.sh:
```diff
@@ -154,6 +154,16 @@ _multioutDocs() {
+    for output in $(getAllOutputNames); do
+        if [ -d "${!output}/share/doc" ]; then
+            while IFS= read -r -d '' dir; do
+                moveToOutput "share/doc/$(basename "$dir")" "${!outputDevdoc}"
```
IIRC, I have seen docs installed to subdirectories so `basename` may not work.

Artturin replied:
so something like this then

```sh
$ export output="/nix/store/something"
$ export dir="/nix/store/something/share/doc/ok"
$ dirname "${dir#"$output/"}"
share/doc

jtojnar commented on pkgs/build-support/setup-hooks/multiple-outputs.sh:
```diff
@@ -154,6 +154,16 @@ _multioutDocs() {
     moveToOutput share/info "${!outputInfo}"
     moveToOutput share/doc "${!outputDoc}"
+
+    # move gi-docgen docs to devdoc
```
```suggestion
    # Move gi-docgen-generated Devhelp docs to devdoc.
```

Artturin committed:
treewide: remove gi-docgen moveToOutput's

these will now be moved in multiple-outputs.sh

we cannot control the output location of gi-docgen

example of how the output location is set
https://github.com/frida/json-glib/blob/a8d3ac569bfaf509e2a20b55ee4fd6b89851b8b1/doc/meson.build#L8

Artturin commented:
tester

```diff
diff --git a/pkgs/development/libraries/json-glib/default.nix b/pkgs/development/libraries/json-glib/default.nix
index efac3690851..40bde74da80 100644
--- a/pkgs/development/libraries/json-glib/default.nix
+++ b/pkgs/development/libraries/json-glib/default.nix
@@ -69,14 +69,17 @@ stdenv.mkDerivation rec {
   doCheck = true;
 
   postFixup = ''
-    # Move developer documentation to devdoc output.
-    # Cannot be in postInstall, otherwise _multioutDocs hook in preFixup will move right back.
-    if [[ -d "$out/share/doc" ]]; then
-        find -L "$out/share/doc" -type f -regex '.*\.devhelp2?' -print0 \
-          | while IFS= read -r -d ''' file; do
-            moveToOutput "$(dirname "''${file/"$out/"/}")" "$devdoc"
-        done
-    fi
+    set -x
+
+    for output in $(getAllOutputNames); do
+        if [ -d "''${!output}/share/doc" ]; then
+            while IFS= read -r -d $'\0' path; do
+                local dirpath="$(dirname "$path")"
+                moveToOutput "''${dirpath#"''${!output}/"}" "''${!outputDevdoc}"
+            done <  <(find "''${!output}/share/doc" -type f -name "*devhelp*" -print0)
+        fi
+    done
+    set +x
   '';
 
   passthru = {
```

```
json-glib> +++++ readlink -m /nix/store/7bbmkfpf1i6c9hcf05xdzwvivzi53kwg-json-glib-1.6.6/share/doc/..
json-glib> ++++ rmdir -p --ignore-fail-on-non-empty /nix/store/7bbmkfpf1i6c9hcf05xdzwvivzi53kwg-json-glib-1.6.6/share
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev = /nix/store/mijb7grfsnxwf4s9m0jhvx0
46kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ local srcPath
json-glib> ++++ for srcPath in "${!output}"/$patt
json-glib> ++++ '[' '!' -e /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev/share/doc/json-glib-1.0 ']'
json-glib> ++++ '[' '!' -L /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev/share/doc/json-glib-1.0 ']'
json-glib> ++++ continue
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc = /nix/store/mijb7grfsnxwf4s9m0jh
vx046kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ continue
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests = /nix/store/mijb7grfsnxw
f4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ local srcPath
json-glib> ++++ for srcPath in "${!output}"/$patt
json-glib> ++++ '[' '!' -e /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests/share/doc/json-gli
b-1.0 ']'
json-glib> ++++ '[' '!' -L /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests/share/doc/json-gli
b-1.0 ']'
json-glib> ++++ continue
json-glib> ++++ IFS=
json-glib> ++++ read -r -d '' path
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' -d /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev/share/doc ']'
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' -d /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc/share/doc ']'
json-glib> ++++ IFS=
json-glib> ++++ read -r -d '' path
json-glib> +++++ find /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc/share/doc -type f -name '*devhe
lp*' -print0
json-glib> +++++ dirname /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc/share/doc/json-glib-1.0/json
-glib-1.0.devhelp2
json-glib> ++++ local dirpath=/nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc/share/doc/json-glib-1.0
json-glib> ++++ moveToOutput share/doc/json-glib-1.0 /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc
json-glib> ++++ local patt=share/doc/json-glib-1.0
json-glib> ++++ local dstOut=/nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc
json-glib> ++++ local output
json-glib> +++++ getAllOutputNames
json-glib> +++++ '[' -n '' ']'
json-glib> +++++ echo 'out dev devdoc installedTests'
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/7bbmkfpf1i6c9hcf05xdzwvivzi53kwg-json-glib-1.6.6 = /nix/store/mijb7grfsnxwf4s9m0jhvx046kg
s5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ local srcPath
json-glib> ++++ for srcPath in "${!output}"/$patt
json-glib> ++++ '[' '!' -e /nix/store/7bbmkfpf1i6c9hcf05xdzwvivzi53kwg-json-glib-1.6.6/share/doc/json-glib-1.0 ']'
json-glib> ++++ '[' '!' -L /nix/store/7bbmkfpf1i6c9hcf05xdzwvivzi53kwg-json-glib-1.6.6/share/doc/json-glib-1.0 ']'
json-glib> ++++ continue
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev = /nix/store/mijb7grfsnxwf4s9m0jhvx0
46kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ local srcPath
json-glib> ++++ for srcPath in "${!output}"/$patt
json-glib> ++++ '[' '!' -e /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev/share/doc/json-glib-1.0 ']'
json-glib> ++++ '[' '!' -L /nix/store/4xjhhssbryvmkasciarr70sq7g3zxw1j-json-glib-1.6.6-dev/share/doc/json-glib-1.0 ']'
json-glib> ++++ continue
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/mijb7grfsnxwf4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc = /nix/store/mijb7grfsnxwf4s9m0jh
vx046kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ continue
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests = /nix/store/mijb7grfsnxw
f4s9m0jhvx046kgs5c3n-json-glib-1.6.6-devdoc ']'
json-glib> ++++ local srcPath
json-glib> ++++ for srcPath in "${!output}"/$patt
json-glib> ++++ '[' '!' -e /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests/share/doc/json-gli
b-1.0 ']'
json-glib> ++++ '[' '!' -L /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests/share/doc/json-gli
b-1.0 ']'
json-glib> ++++ continue
json-glib> ++++ IFS=
json-glib> ++++ read -r -d '' path
json-glib> ++++ for output in $(getAllOutputNames)
json-glib> ++++ '[' -d /nix/store/5r46hdaxy8v1rng5a05fid8bb3yyvdll-json-glib-1.6.6-installedTests/share/doc ']'
json-glib> ++++ set +x
```

SuperSandro2000 commented:
How can we push this forwards?

doronbehar reviewed (commented):
I too think this PR is worth it, and plan to open it again rebased on latest `staging`.
